### PR TITLE
ARROW-9374: [C++][Python] Expose MakeArrayFromScalar

### PR DESCRIPTION
**Pull request by kszucs**
Since we have a complete scalar implementation on the python side, we can implement `pa.repeat(value, size=n)`

### CODE REVIEW

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -297,6 +297,24 @@ def test_nulls(ty):
...
+    (None, 1, pa.array([None])),
+    (None, 10, pa.nulls(10)),
+    (-1, 3, pa.array([-1, -1, -1], type=pa.int64())),
+    (2.71, 2, pa.array([2.71, 2.71], type=pa.float64())),
+    ("string", 4, pa.array(["string"] * 4)),
+    (pa.scalar(8, type=pa.uint8()), 17, pa.array([8] * 17, type=pa.uint8())),
+    (pa.scalar(None), 3, pa.array([None, None, None])),
+    (pa.scalar(True), 11, pa.array([True] * 11)),
+    (False, 9, pa.array([False] * 9)),
+    # ([1, 2], 2, pa.array([[1, 2], [1, 2]]))
```
Comment (kszucs) [inline python/pyarrow/tests/test_array.py:310]: The C++ implementation is not complete, so I'm going to extend its type support.

File: python/pyarrow/array.pxi [hunk 2]
```diff
@@ -334,6 +334,25 @@ def nulls(size, type=None, MemoryPool memory_pool=None):
     return pyarrow_wrap_array(arr)
 
 
+def repeat(value, size, MemoryPool memory_pool=None):
+    cdef:
```
Comment (wesm) [inline python/pyarrow/array.pxi:338]: needs docstring

File: python/pyarrow/tests/test_scalars.py [hunk 3]
```diff
@@ -458,6 +458,7 @@ def test_map():
     assert len(s) == 2
     assert isinstance(s, pa.MapScalar)
     assert isinstance(s.values, pa.Array)
+
```
Comment (wesm) [inline python/pyarrow/tests/test_scalars.py:461]: remove whitespace change

**Review by wesm — COMMENTED** [review]
@kszucs I think you can leave e.g. MapScalar for follow up?

**Comment by kszucs** [comment]
It supports MapScalar, but not the union scalar because we'd need to have the type_code information stored in the union scalar value in order to properly reconstruct a union array if the union type contains identical types.

Created a follow-up JIRA: https://issues.apache.org/jira/browse/ARROW-9434

File: cpp/src/arrow/array/util.cc [hunk 4]
```diff
@@ -323,15 +336,52 @@ class RepeatedArrayFactory {
...
+  }
+
+  Status Visit(const MapType& type) {
+    auto map_scalar = checked_cast<const MapScalar&>(scalar_);
+    auto struct_array = checked_cast<const StructArray*>(map_scalar.value.get());
+
+    ArrayVector keys(length_, struct_array->field(0));
+    ArrayVector values(length_, struct_array->field(1));
+
+    ARROW_ASSIGN_OR_RAISE(auto key_array, Concatenate(keys, pool_));
```
Comment (kszucs) [inline cpp/src/arrow/array/util.cc:375]: We could add a more efficient implementations later.

File: cpp/src/arrow/array/array_test.cc [hunk 5]
```diff
@@ -384,6 +398,40 @@ TEST_F(TestArray, TestMakeArrayFromScalar) {
   }
 }
 
+TEST_F(TestArray, TestMakeArrayFromDictionaryScalar) {
+  auto dictionary = ArrayFromJSON(utf8(), R"(["foo", "bar", "baz"])");
+  auto type = std::make_shared<DictionaryType>(int8(), utf8());
+  ASSERT_OK_AND_ASSIGN(auto value, MakeScalar(int8(), 1));
+  auto scalar = DictionaryScalar({value, dictionary}, type);
```
Comment (wesm) [inline cpp/src/arrow/array/array_test.cc:405]: Note you can also do `DictionaryScalar scalar({..., `

File: cpp/src/arrow/array/array_test.cc [hunk 6]
```diff
@@ -384,6 +398,40 @@ TEST_F(TestArray, TestMakeArrayFromScalar) {
   }
 }
 
+TEST_F(TestArray, TestMakeArrayFromDictionaryScalar) {
+  auto dictionary = ArrayFromJSON(utf8(), R"(["foo", "bar", "baz"])");
+  auto type = std::make_shared<DictionaryType>(int8(), utf8());
```
Comment (wesm) [inline cpp/src/arrow/array/array_test.cc:403]: `::arrow::dictionary` works too

**Review by wesm — APPROVED** [review]
+1, thanks @kszucs

File: cpp/src/arrow/array/array_test.cc (hunk 5 above)
Comment (kszucs) [reply cpp/src/arrow/array/array_test.cc:405]: I know, just my eyes prefer the assignments instead.

**Pull request closed without merging**